cpp/src/arrow/dataset/dataset.cc

```diff
@@ -29,16 +29,13 @@
 namespace arrow {
 namespace dataset {
 
-DataFragment::DataFragment() : scan_options_(ScanOptions::Defaults()) {}
+DataFragment::DataFragment(ScanOptionsPtr scan_options)
+    : scan_options_(std::move(scan_options)), partition_expression_(scalar(true)) {}
 
 SimpleDataFragment::SimpleDataFragment(
     std::vector<std::shared_ptr<RecordBatch>> record_batches, ScanOptionsPtr scan_options)
     : DataFragment(std::move(scan_options)), record_batches_(std::move(record_batches)) {}
 
-SimpleDataFragment::SimpleDataFragment(
-    std::vector<std::shared_ptr<RecordBatch>> record_batches)
-    : SimpleDataFragment(std::move(record_batches), ScanOptions::Defaults()) {}
```

Comment:
To understand this change: this constructor was removed because now ScanOptions always needs to have a schema?

(wondering because it makes using the python bindings a bit more complex. Eg if you have a `FileSystemDataSource`, and you want to get the fragments of this source, you now need to specify a ScanOptions with a schema, while before this was not needed).

Reply:
Yes, this is because schema is a required property of scan options and scan options are a required property of fragments. Datasource is a low-level API and I recommend using the scanner API instead. The scanner builder gets the schema for scan options from a dataset (or a projection thereof) automatically.